### ISSUE TITLE
Web UI login failures with (webauthn / on page refresh)

### ISSUE THREAD

**Issue by yaleman** [kanidm/kanidm#547]
Kanidm webui login hangs when webauthn enabled, then on refresh it gives a sessionid error - or when you just refresh normally?

The sessionid error seems to relate to this: https://github.com/kanidm/kanidm/blob/1de1b2db3bd80a773a74a2f0ced8974a95cb291f/kanidmd/src/lib/event.rs#L699-L707

When a session is restarted it has a complaint about the session ID already existing - the fix might be to just ignore it and set a new one?

## Environment info

* Client: macOS firefox/chrome/safari, current versions. I've tested at <https://webauthn.io/> that the key works in the browsers - it doesn't seem to be related to the key, more the account.
* Kanidmd: built from 7b60d9d03dd7e5c648adc258620ddfedace6c0e5
* User has webauthn and TOTP enabled (disabling webauthn makes it work... sometimes?)

## Steps to reproduce

1. Open new firefox incognito window.
2. Open https://kanidm.example.com
3. Type in username, click begin
4. Username field goes dark, can't do anything
5. Refresh the page, try again, type username, click begin, page reloads and gives the following error:

```## Kanidm Alpha 🦀

An error has occured 😔

missing field `sessionid` at line 1 column 49

8b441a4c-0b01-429c-88e1-022832613371
```

## Network traffic

Looking at network traffic, when I click "Begin" the first time

1. I see a POST to `/v1/auth` with the data: `{"step":{"init":"yaleman"}}`.

Request cookies: `kanidm-session: px7c50a06Agvmtyll7EfY36BYi17Hf3t3WultuHUq4c=JXYECkCrNnrsN9zghSKw82Un6nnEqw0TEfiUIyFMoetUAYTpS+QcjoNiLuK5ect69HY6Iraw8pk5jQG7cPypZw==`

Response is a HTTP/200, Value: `{"sessionid":"00000000-6104-b31b-3110-bc115339c397","state":{"choose":["passwordmfa"]}}`

2. I see a second POST to `/v1/auth` with the data `{"step":{"begin":"passwordmfa"}}`

Request cookies: `kanidm-session: px7c50a06Agvmtyll7EfY36BYi17Hf3t3WultuHUq4c=JXYECkCrNnrsN9zghSKw82Un6nnEqw0TEfiUIyFMoetUAYTpS+QcjoNiLuK5ect69HY6Iraw8pk5jQG7cPypZw==`

Response is a HTTP/200, Value: `{"sessionid":"00000000-6104-b31b-3110-bc115339c397","state":{"continue":["totp",{"webauthn":{"publicKey":{"challenge":"3DrBDdOV1A-ZWJ5n19G7G-u-f3zIoZq2keSx_VTTBws","timeout":60000,"rpId":"kanidm.example.com","allowCredentials":[{"type":"public-key","id":"41g0WriudnddBhUutlJzeOVzxl_WEkIxsw1M6vqE4_xG78FJXauTRTH23mVAuLmouP68tRVpfXFvyEAlJhGAlg"}],"userVerification":"discouraged","extensions":null}}}]}}`

**This is where the first attempt "hangs"**

3. Upon refresh of `/`, the page is loaded with the following cookie:
`"kanidm-session": "px7c50a06Agvmtyll7EfY36BYi17Hf3t3WultuHUq4c=JXYECkCrNnrsN9zghSKw82Un6nnEqw0TEfiUIyFMoetUAYTpS+QcjoNiLuK5ect69HY6Iraw8pk5jQG7cPypZw=="`

4. When clicking begin a request is made to `/v1/auth` with the data `{"step":{"init":"yaleman"}}` which returns a HTTP/500 error. The body of the response is `{"invalidauthstate":"session id present in init"}`

Request cookies: `kanidm-session: px7c50a06Agvmtyll7EfY36BYi17Hf3t3WultuHUq4c=JXYECkCrNnrsN9zghSKw82Un6nnEqw0TEfiUIyFMoetUAYTpS+QcjoNiLuK5ect69HY6Iraw8pk5jQG7cPypZw==`

**Comment by cuberoot74088** [comment]
Hi @yaleman 

I think this bug happens as there is no selection of authsteps implemented. (If I understand this correctly). Can you remove totp and try again? So just one option for MFA.

I had a similar issue when adding backup codes to totp. You can do password but then you would need to choose between backup code and totp.

I think @Firstyear said he wanted to redo the ui and implement that.

Could you check, if this is really the problem? Could you also try to login with the CLI? That should work®  🤠

**Comment by yaleman** [comment]
Ah, that seems to be the case, thanks @cuberoot74088  😄 

The CLI has always worked fine 😄

**Comment by yaleman** [comment]
I'll close this for now as it's waiting on the completion of the login UI

**Comment by Firstyear** [comment]
Thank you @cuberoot74088 for the excellent analysis - you were faster than me to look at this, and completely correct. Thank you!

**Comment by tomberek** [comment]
When doing a quick test I ran into this at the conclusion of a webauthn registration:
```
An Error Occured 🥺
"webauthn"

Operation ID: 96e4d489-bb12-4923-897c-71aa4fefdcfc
```
```
96e4d489-bb12-4923-897c-71aa4fefdcfc 2022-11-06T00:33:26.923445266+00:00 INFO     ┝━    [security.info]: Request received | src: "[::ffff:127.0.0.1]:58396" | http.host: "localhost:8443" | http.method: "POST" | path: "/v1/credential/_update"
96e4d489-bb12-4923-897c-71aa4fefdcfc 2022-11-06T00:33:26.924104047+00:00 TRACE    ┝━ idm_credential_update [ 2.92ms | 80.134% ]
96e4d489-bb12-4923-897c-71aa4fefdcfc 2022-11-06T00:33:26.926491836+00:00 ERROR    │  ┝━ 🚨 [error]: Unable to start passkey registration | e: InvalidRPOrigin
96e4d489-bb12-4923-897c-71aa4fefdcfc 2022-11-06T00:33:26.926519156+00:00 ERROR    │  ┕━ 🚨 [admin.error]: Failed to begin credential_passkey_init | err: Webauthn
96e4d489-bb12-4923-897c-71aa4fefdcfc 2022-11-06T00:33:26.927077907+00:00 ERROR    ┕━ 🚨 [request.error]: Internal error -> Response sent | status: 500 - Internal Server Error
42805fa4-4719-4f6a-85d1-3909dbe3c855 2022-11-06T00:33:26.932729431+00:00 TRACE    tide-request [ 287µs | 100.000% ]
```

**Comment by yaleman** [comment]
From the logs: `Unable to start passkey registration | e: InvalidRPOrigin`

Does the hostname you're connecting to from the web browser match `origin` in your server configuration file?

**Comment by tomberek** [comment]
Yep.... need the port. Thanks! Good work on the UI

**Comment by Firstyear** [comment]
> From the logs: `Unable to start passkey registration | e: InvalidRPOrigin`
> 
> Does the hostname you're connecting to from the web browser match `origin` in your server configuration file?

I think the errors in webauthn have anyhow descriptions, maybe we need to find how to print/ display these?

**Comment by yaleman** [comment]
Yeah, this one looks like the error's not bubbling up correctly - it's currently bailing with a 500 when it could be returning a 400 and presenting it to the user in a nice way.

**Comment by yaleman** [comment]
^ New issue created here

**Comment by Firstyear** [comment]
No I think the 500 is correct, more that when we display the webauthn error in the log, there is a long text form we can add too.
